Userland/Libraries/LibCrypto/BigInt/Algorithms/ModularPower.cpp

```diff
@@ -49,4 +49,239 @@ void UnsignedBigIntegerAlgorithms::destructive_modular_power_without_allocation(
+ */
+UnsignedBigInteger::Word UnsignedBigIntegerAlgorithms::montgomery_fragment(UnsignedBigInteger& z, size_t offset_in_z, UnsignedBigInteger const& x, UnsignedBigInteger::Word y_digit, size_t num_words)
+{
+    UnsignedBigInteger::Word carry { 0 };
+    for (size_t i = 0; i < num_words; ++i) {
+        UnsignedBigInteger::Word a_carry;
+        UnsignedBigInteger::Word a;
+        linear_multiplication_with_carry(x.m_words[i], y_digit, z.m_words[offset_in_z + i], a_carry, a);
+        UnsignedBigInteger::Word b_carry;
+        UnsignedBigInteger::Word b;
+        addition_with_carry(a, carry, b_carry, b);
+        z.m_words[offset_in_z + i] = b;
+        carry = a_carry + b_carry;
+    }
+    return carry;
+}
+
+/**
+ * Computes the "almost montgomery" product : x * y * 2 ^ (-num_words * BITS_IN_WORD) % modulo
+ * [Note : that means that the result z satisfies z * 2^(num_words * BITS_IN_WORD) % modulo = x * y % modulo]
+ * assuming :
+ *  - x, y and modulo are all already padded to num_words
+ *  - k = inverse_wrapped(modulo) (optimization to not recompute K each time)
+ * Algorithm from: Gueron, "Efficient Software Implementations of Modular Exponentiation". (https://eprint.iacr.org/2011/239.pdf)
+ */
+void UnsignedBigIntegerAlgorithms::almost_montgomery_multiplication_without_allocation(
+    UnsignedBigInteger const& x,
+    UnsignedBigInteger const& y,
+    UnsignedBigInteger const& modulo,
+    UnsignedBigInteger& z,
+    UnsignedBigInteger::Word k,
+    size_t num_words,
+    UnsignedBigInteger& result)
+{
+    VERIFY(x.length() >= num_words);
+    VERIFY(y.length() >= num_words);
+    VERIFY(modulo.length() >= num_words);
+
+    z.set_to(0);
+    z.resize_with_leading_zeros(num_words * 2);
+
+    UnsignedBigInteger::Word previous_double_carry { 0 };
+    for (size_t i = 0; i < num_words; ++i) {
+        // z[i->num_words+i] += x * y_i
+        UnsignedBigInteger::Word carry_1 = montgomery_fragment(z, i, x, y.m_words[i], num_words);
+        // z[i->num_words+i] += modulo * (z_i * k)
+        UnsignedBigInteger::Word t = z.m_words[i] * k;
+        UnsignedBigInteger::Word carry_2 = montgomery_fragment(z, i, modulo, t, num_words);
+
+        // Compute the carry by combining all of the carrys of the previous computations
+        // Put it "right after" the range that we computed above
+        UnsignedBigInteger::Word temp_carry = previous_double_carry + carry_1;
+        UnsignedBigInteger::Word overall_carry = temp_carry + carry_2;
+        z.m_words[num_words + i] = overall_carry;
+
+        // Detect if there was a "double carry" for this word by checking if our carry results are smaller than their components
+        previous_double_carry = (temp_carry < carry_1 || overall_carry < carry_2) ? 1 : 0;
+    }
+
+    if (previous_double_carry == 0) {
+        // Return the top num_words bytes of Z, which contains our result.
+        shift_right_by_n_words(z, num_words, result);
+        result.resize_with_leading_zeros(num_words);
```

Comment:
Can this ever end up with less than `num_words` words?
`shift_right_by_n_words` should be allocating resizing to `num_words`, as far as I can see.

Reply:
At the moment, no. But this is an implementation details of the shift_right operator. If we modify this operator later to normalize the BigInt, then it could remove leading zeros.
Since the resize_with_leading_zeros() operation is a no-op, it felt appropriate to call it explicitly.

Reply:
Alright, seems reasonable :P